pkg/services/ngalert/notifier/images.go

```diff
@@ -44,3 +51,101 @@ func (i imageStore) GetImage(ctx context.Context, uri string) (*images.Image, er
 		CreatedAt: image.CreatedAt,
 	}, nil
 }
+
+func (i imageProvider) GetImageURL(ctx context.Context, alert *alertingNotify.Alert) (string, error) {
+	uri, err := getImageURI(alert)
+	if err != nil {
+		return "", err
+	}
+
+	// If the identifier is a URL, validate that it corresponds to a stored, non-expired image.
+	if strings.HasPrefix(uri, "http") {
+		exists, err := i.store.URLExists(ctx, uri)
+		if err != nil {
+			return "", err
+		}
+		if !exists {
+			return "", images.ErrImageNotFound
+		}
+		return uri, nil
+	}
+
+	// If the identifier is a token, remove the prefix, get the image and return the URL.
+	token := strings.TrimPrefix(uri, "token://")
+	return i.getImageURLFromToken(ctx, token)
+}
+
+// getImageURLFromToken takes a token and returns the URL of the image that token belongs to.
+func (i imageProvider) getImageURLFromToken(ctx context.Context, token string) (string, error) {
+	image, err := i.store.GetImage(ctx, token)
+	if err != nil {
+		if errors.Is(err, models.ErrImageNotFound) {
+			return "", images.ErrImageNotFound
+		}
+		return "", err
+	}
+
+	if !image.HasURL() {
+		return "", images.ErrImagesNoURL
+	}
+	return image.URL, nil
+}
+
+func (i imageProvider) GetRawImage(ctx context.Context, alert *alertingNotify.Alert) (io.ReadCloser, string, error) {
+	uri, err := getImageURI(alert)
+	if err != nil {
+		return nil, "", err
+	}
+
+	// Check whether the uri is a URL or a token to know how to query the DB.
+	var image *models.Image
+	if strings.HasPrefix(uri, "http") {
+		image, err = i.store.GetImageByURL(ctx, uri)
+	} else {
+		token := strings.TrimPrefix(uri, "token://")
+		image, err = i.store.GetImage(ctx, token)
+	}
+	if err != nil {
+		if errors.Is(err, models.ErrImageNotFound) {
+			return nil, "", images.ErrImageNotFound
+		}
+		return nil, "", err
+	}
+
+	if !image.HasPath() {
+		return nil, "", images.ErrImagesNoPath
+	}
+
+	// Return image bytes and filename.
+	readCloser, err := openImage(image.Path)
+	if err != nil {
+		return nil, "", err
+	}
+	filename := filepath.Base(image.Path)
+	return readCloser, filename, nil
+}
+
+// getImageURI is a helper function to retrieve the image URI from the alert annotations as a string.
+func getImageURI(alert *alertingNotify.Alert) (string, error) {
+	uri, ok := alert.Annotations[alertingModels.ImageTokenAnnotation]
+	if !ok {
+		return "", fmt.Errorf("no image uri in annotations")
```

Comment:
I think it should return ErrImageNotFound. The absence of the token annotation is normal situration

Reply:
Returning `ErrImageNotFound` would imply that we have a URI but when we query the database we can't find any non-expired images for that URI. This is a different scenario, we're checking for a URI in the annotations but there's none.
Since it's a normal situation I don't see the need for returning a named error. In any case, I would consider not returning any error at all.

Reply:
I'll use `images.NoImageForAlert`